src/pydicom/encaps.py

```diff
@@ -1,509 +1,624 @@
-        All fragments as a list of ``bytes``.
-    """
-    # Convert data into a memory-mapped file
-    with DicomBytesIO(data) as fp:
-        # DICOM standard requires this
-        fp.is_little_endian = True
-        BasicOffsetTable = read_item(fp)  # NOQA
-        seq = []
+    # nr_fragments < number_of_frames
+    raise ValueError(
+        "Unable to generate frames from the encapsulated pixel data as there "
+        "are fewer fragments than frames; the dataset may be corrupt or the "
+        "number of frames may be incorrect"
+    )
 
-        while True:
-            item = read_item(fp)
 
-            # None is returned if get to Sequence Delimiter
-            if not item:
-                break
-            seq.append(item)
+def generate_frames(
+    buffer: bytes | ReadableBuffer,
+    /,
+    *,
+    number_of_frames: int | None = None,
+    extended_offsets: tuple[list[int], list[int]] | tuple[bytes, bytes] | None = None,
+    endianness: str = "<",
+) -> Iterator[bytes]:
+    """Yield complete pixel data frames from `buffer`.
 
-        # XXX should
-        return seq
+    .. versionadded:: 3.0
 
+    .. note::
 
-def defragment_data(data: bytes) -> bytes:
-    """Read encapsulated data and return the fragments as one continuous bytes.
+        When the Basic Offset Table is empty and the Extended Offset Table
+        isn't supplied then more frames may be yielded than given by
+        `number_of_frames` provided there are sufficient excess fragments
+        available.
 
     Parameters
     ----------
-    data : bytes
-        The encapsulated pixel data fragments.
+    buffer : bytes | readable buffer
+        A buffer containing the encapsulated frame data, starting at the first
+        byte of the basic offset table. May be :class:`bytes` or an object
+        with ``read()``, ``tell()`` and ``seek()`` methods. If the latter
+        then the final offset position depends on the number of yielded frames.
+    number_of_frames : int, optional
+        Required for multi-frame data when the Basic Offset Table is empty,
+        the Extended Offset Table has not been supplied and there are
+        multiple frames. This should be the value of (0028,0008) *Number of
+        Frames* or the expected number of frames in the encapsulated data.
+    extended_offsets : tuple[list[int], list[int]] or tuple[bytes, bytes], optional
+        The (offsets, lengths) of the Extended Offset Table as taken from
+        (7FE0,0001) *Extended Offset Table* and (7FE0,0002) *Extended Offset
+        Table Lengths* as either the raw encoded values or a list of their
+        decoded equivalents.
+    endianness : str, optional
+        If ``"<"`` (default) then the encapsulated data uses little endian
+        encoding, otherwise if ``">"`` it uses big endian encoding.
 
-    Returns
-    -------
+    Yields
+    ------
     bytes
-        All fragments concatenated together.
+        A single frame of encoded pixel data.
```

Comment:
Something like "All frames of encoded pixel data, one frame at a time"?

Reply:
Done